### DISCUSSION TITLE
Error compiling 'cart_reader' v9.5- 'SdFat.h: No such file or directory'

### DISCUSSION THREAD

**Discussion by benmerrick3d** [sanni/cartreader#498] [Need help]
Hi,

I am trying to update my cart reader, which is version 3 pre-built by save the hero builders: https://savethehero.builders/collections/build-service/products/open-source-cartridge-reader-v3-alter-build-service

I am following this and have got to step 9: https://github.com/sanni/cartreader/wiki/How-to-flash-the-Arduino

The only things I have changed in my code are lines 118 and 124 to disable clock gen, and 71 to enable HW3.

These are the changes I have made:

Starting line 65
`'//****************************************** // !!! CHOOSE HARDWARE VERSION !!! //****************************************** // Remove // in front of the line with your hardware version // #define HW5 // #define HW4 #define HW3 // #define HW2 // #define HW1 // #define SERIAL_MONITOR`'

Starting line: 115
```
#if (defined(HW2) || defined(HW3))
#define enable_OLED
#define enable_Button2
//#define clockgen_installed
//#define fastcrc
#endif

#if defined(HW1)
#define enable_OLED
// #define clockgen_installed
// #define fastcrc
#endif

```

When I click to either 'Verify/Compile' or 'Upload' I get this error:

`Cart_Reader:178:10: fatal error: SdFat.h: No such file or directory #include "SdFat.h" ^~~~~~~~~ compilation terminated. exit status 1 SdFat.h: No such file or directory`

Does anybody know why I am getting this error or what I can do to get round it?

I have looked through the folders and found the SdFat.h file in here, I assume this is the correct place? (I just unzipped to desktop)

C:\Users\Local Admin\Desktop\V9.5_Portable\Arduino IDE Portable\portable\sketchbook\libraries\SdFat\src
![image](https://user-images.githubusercontent.com/30409503/187091103-c99c95e4-be55-4ac5-8ab9-2408f09bc4cc.png)

I have also just tried v9.4 and I get the same error, so I assume its a mistake my end, do you have any idea what I am doing wrong? 

Thanks

**Comment by benmerrick3d** [comment]
I also just tried v9.1 and v8.4 with the only change being to remove // for 
`#define HW3`

Same error

**Comment by sanni** [comment]
You can download and install https://downloads.arduino.cc/arduino-1.8.19-windows.exe and then copy the "libraries" and "Cart_Reader" folder from "V9.5_Portable\Arduino IDE Portable\portable\sketchbook\" to "C:\Users\your_username\Documents\Arduino\" and launch the Arduino IDE from the start menu.

**Reply by benmerrick3d** [reply]
Thank you, that works.

I don't understand the extra options in step 7, what do they do and why when should I enable them? 
![image](https://user-images.githubusercontent.com/30409503/187211246-85f519ee-3ed7-47da-81ea-4fba927d251e.png)

Also to double check my understanding- I am having issues getting eep save files off N64 cartridges, I don't have adafruit clockgen chip installed, I believe this is a requiement and I will need to install that to work?

I am also having issues with mpk files from n64 memory cards, the files seem to be corrup or unusable. I see in the latest update there was an improvement linked to that, could that help? Do you know an easy way to verify the integrity of mpk files?

Thank you

**Comment by sanni** [comment]
You don't need the Clockgen, just remove the line "#define clockgen_installed" and then set the switches like so:
![68747470733a2f2f646c2e64726f70626f7875736572636f6e74656e742e636f6d2f732f366275666279326a786d3537756a6d2f6e36345f6a756d7065723376332e6a70673f646c3d31](https://user-images.githubusercontent.com/19169494/187212948-a8843ba4-49bd-4e62-bb2e-4e4965fab4a8.jpg)

enable_PCE is for dumping PC Engine, enable_WS for Wonderswan, enable_NGP for Neogeo Pocket. If you don't have those systems just ignore the extra step.

You can upload mpk files to this site to see if they are ok: https://bryc.github.io/mempak/

You can also flash V5.5_Portable from [here](https://github.com/sanni/cartreader/wiki/Cart-Reader-V3#older-releasesfor-testing-newer-releases-will-work-with-v3-hardware-too) to see if that works better with dumping MPK. It's the last version with the old SD card library.

**Reply by benmerrick3d** [reply]
So I should be able to get the rom and the saves from N64 carts that have 4Kbit eeprom and 16Kbit eeprom save types, for example these on this list? http://micro-64.com/database/gamesave.shtml

When I have tried sometimes it hangs on getting the save 'Reading eeprom' (left it for 40mins and still no progress). I have managed to get a couple of games to work. 

The 3 that don't work are star wars episode 1 racer, goldeneye and perfect dark- they all have 4kbit or 16kbit eeprom saves.

I tried on whatever version came on the cart reader when I bought it and now just tried with v9.5 (clockgen \\ out).

As I mentioned above, I bought the cart reader but i dont have the adafruit, should it work? Any ideas I can try? I'm desperate to get my saves off before the batteries die. Thanks for your help

**Reply by benmerrick3d** [reply]
Just to add, I've tried all 3 cartridges in my N64 console and all of the save data shows first time, no issues so the batteries are still alive.

If for some reason it's actually doing something but just being extremely slow, how long is it safe to leave it on the 'Reading eep...' screen? I'm worried leaving it for hours on end would cause harm to the device or cartridge. 

I'll try the older versions for the mpk files once I get these saves done. I tried an online tester for the mpk files I managed to get but they didn't load in that website sadly. Thanks

**Reply by SAVE-THE-HERO** [reply]
Hello benmerrick3d
I am a seller for save the hero builders
The clock generator does not have to be adafruit, but it has been tested and functions the same. If you have our tested clock generator.
I tested and needed a clock generator to save the eeprom for just in our setting.
So pls one try, clock generator setting to On.
![FYPbtgIaUAA_Kve](https://user-images.githubusercontent.com/89830526/187604405-b0e39a9b-c324-4c89-a76e-11f7ee8def5e.jpg)
If the contact is working well, it will not freeze.
Saving the save data is instantaneous.
Complete contact is of utmost importance.


One thing to keep in mind when you  damp.
https://github.com/sanni/cartreader/discussions/434
https://twitter.com/suneon1/status/1550324277479878656?s=20&t=_CrnvF5jT5LRfwR7CKLk-Q

and you can ask to my mail
savetheherobuilders@gmail.com
I can answer your questions.
but I do not know about mpk.
Thank you

**Reply by benmerrick3d** [reply]
I did some more tests and the clock gen wasn't needed- once I commented out the clock gen part of the code. 
The thing I had to do with N64 carts is try a LOT of times to get it to work, including putting the cart in on the right side of the socket and sliding it over to the left side. Sometimes that would work, sometimes not. After about 20 or so tries on each cartridge it would work one time

**Reply by benmerrick3d** [reply]
I haven't yet managed to get a successful mpk file though, I will have to try going back to v5 as suggested above

**Reply by benmerrick3d** [reply]
@sanni thanks for your help before- I now have the chance to try the v5.5
I've loaded it in Arduino, what do I need to comment out as I dont have a adafruit clockgen in this code?
I've found this at 97:
```
// Adafruit Clock Generator
#include <si5351.h>
Si5351 clockgen;

```
Should I change it to:
```
// Adafruit Clock Generator
//#include <si5351.h>
//Si5351 clockgen;
```


When I do that I get this error:
![image](https://user-images.githubusercontent.com/30409503/188119392-06ae2837-2fb6-4ab1-a15b-295821421c71.png)

**Reply by sanni** [reply]
You don't have to edit anything in the code. It's an older version that was made just for HW3. before HW4 and HW5 were released.

You will have to delete the "libraries" and "Cart_Reader" folder in "C:\Users\your_username\Documents\Arduino" that you copied from the 9.5 release and copy the identically named folders from "V5.5_Portable\Arduino IDE Portable\portable\sketchbook" to "C:\Users\your_username\Documents\Arduino".

**Reply by benmerrick3d** [reply]
Okay thanks, I did that and uploaded. Grabbed the mpk off the n64 controllers and unfortunately they still give files that wont open in https://bryc.github.io/mempak/

I tried using them (assuming they are fine) but it kept showing issues with the files. I opened it in a hex editor and it has data in it (not just all 0s).

Have you successfully tried coping a mpk from an n64 controller?

**Reply by SAVE-THE-HERO** [reply]
We got mpk file.
You can see this
https://www.youtube.com/watch?v=C0NlGYy_RGg
Can you test N64 controller?
<img width="218" alt="スクリーンショット 2022-09-02 20 56 57" src="https://user-images.githubusercontent.com/89830526/188148483-d356622f-cfd4-44b4-9e96-85e08246bd8c.png">
<img width="845" alt="スクリーンショット 2022-09-02 20 55 43" src="https://user-images.githubusercontent.com/89830526/188148487-1afdbac5-cd09-41dc-81f1-1f5ac83576dd.png">

**Reply by sanni** [reply]
> Grabbed the mpk off the n64 controllers 

Are you using an original N64 controller made by Nintendo or some 3rd Party controller?
Can you upload the mpk here? To what game does the save belong or are there multiple game saves on the mpk?

**Reply by benmerrick3d** [reply]
I'm using official controllers. I've done the test on the controllers and they work fine.

Here is an upload of the files- some are from a memory pack I believe has corrupted, but would be good to know for sure. Others the pack worked when in the N64- 177 and 214 should be safe ones. I have also shared the original mpk files, and some which have been converted to .n64 and .srm using converters online to see if I could make any progress
[mpkfiles.zip](https://github.com/sanni/cartreader/files/9479386/mpkfiles.zip)

**Reply by benmerrick3d** [reply]
Hey @sanni I just wondered if you had a chance to look at the files? Thank you for your help, I appreciate it

**Reply by sanni** [reply]
I could not get any info out of the dumped files but I have added some error checking to the code so at least you can now find out why the reading fails. You can download the new version [here](https://github.com/sanni/cartreader/archive/refs/heads/master.zip). It uses the same Arduino libraries as the 9.5 portable release.

One thing I did notice was that the Cart Reader dumps the Controller Pak pretty fast compared to e.g. a Raphnet USB adapter, so what you can try is adding a bigger delay between blocks by changing this number in [line L2136 of N64.ino](https://github.com/sanni/cartreader/blob/master/Cart_Reader/N64.ino#L2136) from 800 to maybe 2000.

````
  // Real N64 has about 627us pause between banks, add a bit extra delay
      if (currBlock < 479)
        delayMicroseconds(800);
    }
````

I only have an original/unmodded Controller Pak so I don't know if slower Controller Paks exist.

**Reply by benmerrick3d** [reply]
Thank you. I copied 'Cart_reader' from the download link into 'D:\Users\Name\Documents\Arduino' and replaced the files- I tried with the v9.5 libraries but get this error:

![image](https://user-images.githubusercontent.com/30409503/189431680-32baca58-b886-45ba-840c-15937b35d42d.png)

**Reply by sanni** [reply]
Delete the folder libraries in D:\Users\Name\Documents\Arduino\ and copy the libraries folder from V9.5_Portable\Arduino IDE Portable\portable\sketchbook to D:\Users\Name\Documents\Arduino\.

**Reply by benmerrick3d** [reply]
Sorry I think I copied the wrong libraries, didnt do portable. 1 minute

**Reply by benmerrick3d** [reply]
Yeah my mistake, I didnt realise there were 2 libraries folders. Just uploading now, will test with the controllers in 5mins

**Reply by benmerrick3d** [reply]
Edit: ignore this message, updated in another reply with newer memory pack files

```
One of the memory packs didnt read as it said 'could not be found'- I think that one might be dead.

The other one copied these 2 files, but it gave 2 errors: 'header' and 'toc'

Is that a problem with the memory pack or is that a reader software problem?

Thanks
```

**Reply by benmerrick3d** [reply]
[MPK_headerandtocerror.zip](https://github.com/sanni/cartreader/files/9538362/MPK_headerandtocerror.zip)

I managed to get files off both memory packs (one controler is more reliable). 215 and 216 are the same memory pack, one with eep switch on and one off (not sure what its meant to be for memory packs). 217 and 218 are the other memory pack, one with eep one with not.

All 4 readings gave the errors: 'header' and 'toc'

Is that a problem with both memory packs or is that a reader software problem?

**Reply by sanni** [reply]
From what game is the save on the Controller Pak? Maybe I can try recreating the situation.

**Reply by benmerrick3d** [reply]
I can't check on console until tomorrow but pretty certain it's Perfect Dark on 215 and 216- it has that written on it too. That's the main game I used my memory packs for

**Reply by sanni** [reply]
I have added another safety check and fixed a bug. Could you try again: https://github.com/sanni/cartreader/archive/refs/heads/master.zip

**Reply by benmerrick3d** [reply]
Just before I try that, I checked the memory paks in the manager in the perfect dark game (hold start when you turn the console on) and did a repair on both memory paks.

Both show up in the N64 fine as perfect dark saves. 
Memory pak 1: 221 and 222 (one is with eep on one without) show as vaild on this website: https://bryc.github.io/mempak/  and they also dont show errors on the cart reader

Memory pak 2: 219 and 220 (one is with eep on one without) dont show as valid on the website and both error as: 'header' and 'toc'

[MPK_220910.zip](https://github.com/sanni/cartreader/files/9540596/MPK_220910.zip)


I'll try the new version now for you

**Reply by benmerrick3d** [reply]
These are the files after bug fix/latest check you just asked for:

[MPK_221009_bugfix.zip](https://github.com/sanni/cartreader/files/9540636/MPK_221009_bugfix.zip)

Memory pak 1: header 0/4 and toc 0/2 saves: 223 and 224

Memory pak 2: header 4/4 and toc 2/2 saves: 225 and 226

How do I use a .crc file? It outputted those this time too

**Reply by benmerrick3d** [reply]
Also do you know how actually use these mpk files in something like retroarch? I am trying to test if the memory pak that seems okay actually works on the PC.

Thank you for your help

**Reply by sanni** [reply]
The CRC file is just for the Cart Reader to validate the dumped mpk file in the SD. 
It could be that the one Controller Pak simply ran out of battery. They have a coin cell inside that needs to be replaced every ~15 years.
